Added the ec2_run policy

athiruma opened the pull request:
## Type of change
Note: Fill **x** in []
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] dependencies

## Description
<!--- Describe your changes below -->
Added the ec2_run policy.
Created the ADR for the changes

## For security reasons, all pull requests need to be approved first before running any automated CI

ebattat commented:
The best practice is to split this PR to 2:
1. Add ADR
2. Add ec2 run policy
When you are mixing between 2 functions its difficult to debug it in case of errors and also difficult to review it.

ebattat commented on cloud_governance/common/helpers/aws/aws_cleanup_operations.py:
```diff
@@ -0,0 +1,153 @@
+        :rtype:
+        """
+        action = "deleted"
+        try:
```
Should replace to case

athiruma replied:
I wonder, does python have the switch case? 🤔

ebattat commented on cloud_governance/common/helpers/aws/aws_cleanup_operations.py:
```diff
@@ -0,0 +1,153 @@
+                    return tag.get('Value').strip()
+        return ''
+
+    def get_clean_up_days_count(self, tags: list):
```
Seems it also updates the cleanup days value ?

athiruma replied:
No, It returns the clean_up days count. We have another method which updates the days count.
[Check here](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit).

ebattat commented on cloud_governance/common/helpers/aws/aws_cleanup_operations.py:
```diff
@@ -0,0 +1,153 @@
+                self._s3_client.put_bucket_tagging(Bucket=resource_id, Tagging={'TagSet': tags})
+            elif self._policy == 'empty_roles':
+                self._iam_client.tag_role(RoleName=resource_id, Tags=tags)
+            elif self._policy in ('ip_unattached', 'unused_nat_gateway', 'zombie_snapshots', 'ebs_unattached', 'ec2_run'):
```
What about new policies that we will add in the future line s3_age, do you need add it also here ?
its very risky to tide to specific policy name !!!

athiruma replied:
Yes, I know. I was thinking about how I change the process. As of now, we will stick to it later we can replace it.

ebattat commented on cloud_governance/main/aws_main_operations.py:
```diff
@@ -0,0 +1,41 @@
+        policies_list = self.__get_policies()
+        for policy_type, policies in policies_list.items():
+            # @Todo support for all the aws policies, currently supports ec2_run as urgent requirement
+            if self.__policy in policies and self.__policy == "ec2_run":
```
Why you ask for specific policy  self.__policy == "ec2_run"

athiruma replied:
As I specified in the new ADR, this is only for ec2_run. So this should work against only ec2_run. I'll activate other policies later. Its risky to activate all policies without testing.

ebattat commented on cloud_governance/main/environment_variables.py:
```diff
@@ -48,7 +52,7 @@ def __init__(self):
         self._environment_variables_dict['policy'] = EnvironmentVariables.get_env('policy', '')
 
-        self._environment_variables_dict['aws_non_cluster_policies'] = ['ec2_idle', 'ec2_stop', 'ec2_run', 'ebs_in_use',
+        self._environment_variables_dict['aws_non_cluster_policies'] = ['ec2_idle', 'ec2_stop', 'ebs_in_use',
```
Why do u remove ec2_run from this list ?

athiruma replied:
It is the static approach. It is very bad to have like this.
Look this new [ADR Dir structure](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit). I added **policy/cleanup**, so all cleanup policies will reside in this folder and we will walk through this and run the policies.

ebattat commented on cloud_governance/policy/aws/cleanup/ec2_run.py:
```diff
@@ -0,0 +1,84 @@
+            else:
+                cleanup_days = 0
+                force_tag_update = self._aws_cleanup_policies.get_force_tag_update()
+            self._aws_cleanup_policies.update_resource_tags(resource_id=instance.get('InstanceId'),
```
I think the name should be update_resource_day_counts and not all the tags, do you update only the day count ?

athiruma replied:
Yes, I can rename the method. 👍

ebattat commented on cloud_governance/policy/policy_operations/aws/zombie_non_cluster/zombie_non_cluster_polices.py:
```diff
@@ -22,32 +22,31 @@ def run(self):
+
+                    if self._es_operations.check_elastic_search_connection():
+                        if policy_result:
+                            if len(policy_result) > 500:
```
can you update what the meaning of 500 ?

ebattat commented on docs/source/index.md:
```diff
@@ -11,7 +11,7 @@ This tool support the following policies:
 * Real time Openshift Cluster cost, User cost
 * [ec2_idle](../../cloud_governance/policy/aws/ec2_idle.py): idle ec2 in last 4 days, cpu < 2% & network < 5mb.
-* [ec2_run](../../cloud_governance/policy/aws/ec2_run.py): running ec2.
+* [ec2_run](../../cloud_governance/policy/aws/cleanup/ec2_run.py): running ec2.
```
Why you add cleanup ?

athiruma replied:
I moved ec2_run to the cleanup folder. So we have all cleanup policies collectively. [ADR](https://docs.google.com/document/d/1hPv80Xe6-Cao59smi3Yougsa9lfhYparpxVTmNpIM94/edit)

athiruma commented:
> The best practice is to split this PR to 2:
> 
> 1. Add ADR
> 2. Add ec2 run policy
>    When you are mixing between 2 functions its difficult to debug it in case of errors and also difficult to review it.

I added ADR for the ec2_run policy. It's not the two separate PRs. It was the better time to have the ADR.  Thats why I added the major change.
